### ISSUE TITLE
Contrast experiment

### ISSUE THREAD

**Issue by puhan123** [ZiweiWangTHU/BiDet#20]
When you do a comparison experiments, For example, Bi-Real method is used to binarize the network,.The SSD_vgg16 structure used in Bi-Real is same as the SSD_vgg16 used in BiDet(SC)? Structure changes include add BN, clip maxpool,and so on.

**Comment by Wuziyi616** [comment]
Yes. We perform fair comparison to set every variables, settings and network architectures the same. Only the loss function is different. And we make sure both models are properly trained to convergence by several runs.

**Comment by puhan123** [comment]
Thank you for your reply!

When you use Bi-Real to binary the SSD_VGG, did you adopt the   Bi-Real Net's training strategy(first real-valued with relu activation, then clipping activation, finally   binary activation & weights)?

**Comment by Wuziyi616** [comment]
No, we didn't use it. Because in the later version of Bi-Real-Net, the authors pointed out the it is unnecessary to do so. Directly training the binary values from scratch can get good results ([here](https://github.com/liuzechun/Bi-Real-net) README file, Pytorch Implementation section). We just follow their words because I think applying its original training strategy to a detection training is too slow. But I have to admit that this may not be optimal for binary detection, and there are indeed BNNs papers show that step-by-step binarization and training can benefit binary detectors. For example, [this paper](https://www.sciencedirect.com/science/article/pii/S0167865519301096) adopts a greedy layer-wise binarization schedule. Maybe you can try it by yourself :)

**Comment by puhan123** [comment]
Thank you very much. It helped me a lot!

**Comment by puhan123** [comment]
Hello! According to Bi-Det, I modify the ssd in MMdetection  v2 .I try to binarize SSD_vgg16 used Bi-Real. The SSD_vgg16 structure used in Bi-Real is same as the SSD_vgg16 used in BiDet(SC). But I found that the network does not converge.

I don't know what to do next?

**Comment by Wuziyi616** [comment]
First I want to make sure that, you use the Bi-Real Conv modules provided in my repo? Also, you need to make sure that the first and last (output) layers of your SSD_VGG16 is not binarized. And do you use the pretrained Bi-Real-VGG weight I provide? Do you load that weight successfully?

If these are true, please report your training settings, e.g. dataset, batchsize, optimizer, learning rate. If possible, give me you training loss curve. It's hard to find out what's happening given your description of "does not converge".

**Comment by puhan123** [comment]
Yes，I use the Bi-Real Conv modules provided in my repo. And the first and last(output) layers of my SSD_VGG16 is not binarized. I also use the pretrained Bi-Real-VGG you provide. And I load that weight succesfully. 
The main code of loading weight is like below: 
_if isinstance(pretrained, str):
            logger = get_root_logger()
            load_checkpoint(self.vgg_layers, pretrained, strict=True, logger=logger)_


**My training settings are as below:**
**dataset**: VOC(train:VOC2007+VOC2012, test:VOC2007)
**batchsize**:20  (samples_per_gpu=20)
**the num of gpu**:2
**optimizer**:SGD  optimizer = dict(type='SGD', lr=1e-5, momentum=0.9, weight_decay=5e-4)
**learning rate**:1e-5
**learning policy**:
lr_config = dict(
    policy='CosineAnnealing',
    by_epoch=True,
    min_lr_ratio=0,
    warmup_iters=500,
    warmup_ratio=0.9)

em, I'm trying to plot the training loss curve ,and I will update later.

Thank you very much!

**Comment by Wuziyi616** [comment]
I think maybe I know what's wrong here. You see, I use init_lr=1e-3 in my repo as [here](https://github.com/ZiweiWangTHU/BiDet/blob/master/ssd/train_bidet_ssd.py#L65), and 1e-5 is definitely too small for initialization. Besides, a very important lesson I learned in BNNs is that Adam optimizer is superior to SGD in most cases. If the error you encountered is not loss explosion (which means loss goes to very big values and then NaN), I'd suggest you to use Adam with initial_lr=1e-3 (or 1e-4) and have a try. Hope this will help you.

**Comment by puhan123** [comment]
Thank you very much. You are so nice. I will try it.

**Comment by puhan123** [comment]
Hello! I found my network is converge now. Thank you very much again!
I would like to confirm one more thing with you .
I set optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2)), Is it OK? I see you set GRADIENT_CLIP_NORM = 5.

**Comment by Wuziyi616** [comment]
Yes, I do think that you need to perform gradient clipping if you use VGG as the backbone, whereas it is not necessary for ResNet backbone. However, I am not sure why you use 35? Maybe it is also OK but I guess 35 is quite a large magnitude?

**Comment by puhan123** [comment]
Thank you very much! You are so nice! I will do a test for it .

**Comment by puhan123** [comment]
Hello! I am the same as you to binarize the SSD_VGG16(Using Bi-Real method). And I trained 200 epoch, and I found its mAP==45.8. There is still a big gap for your paper (63.8). What can I do next? Thank you.

**Comment by Wuziyi616** [comment]
Humm, that's quite a low mAP. I'd suggest you to try training for 100 epochs and see whether you get a higher or lower mAP. From my experience, 200 epochs on VOC may cause severe over-fitting (but I'm not sure, because in my experiment I use the number of iterations instead of epoch to determine when to decay lr, but epoch should also be OK).

**Comment by puhan123** [comment]
Thank you very much! You are so nice! I will try it .

**Comment by kj-kx** [comment]
hello,when i use the bidet ,it takes several days to converge，i wonder know if the use of IB principle limits the speed，i also want try to imporve the code from a more simple form.could you offer me the code(Using Bi-Real method)?thank you！
